plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/NessieIcebergClient.java

```diff
@@ -0,0 +1,432 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.nessie;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.plugin.hive.SchemaAlreadyExistsException;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.SchemaNotFoundException;
+import io.trino.spi.connector.SchemaTableName;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.util.Tasks;
+import org.projectnessie.client.api.CommitMultipleOperationsBuilder;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.client.http.HttpClientException;
+import org.projectnessie.error.BaseNessieClientServerException;
+import org.projectnessie.error.NessieConflictException;
+import org.projectnessie.error.NessieNamespaceAlreadyExistsException;
+import org.projectnessie.error.NessieNamespaceNotEmptyException;
+import org.projectnessie.error.NessieNamespaceNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.error.NessieReferenceNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.EntriesResponse;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.ImmutableIcebergTable;
+import org.projectnessie.model.Namespace;
+import org.projectnessie.model.Operation;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.Tag;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.iceberg.IcebergErrorCode.ICEBERG_CATALOG_ERROR;
+import static io.trino.plugin.iceberg.IcebergErrorCode.ICEBERG_COMMIT_ERROR;
+import static io.trino.plugin.iceberg.catalog.nessie.NessieIcebergUtil.buildCommitMeta;
+import static io.trino.plugin.iceberg.catalog.nessie.NessieIcebergUtil.toKey;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class NessieIcebergClient
+{
+    private final NessieApiV1 nessieApi;
+    private final Supplier<UpdateableReference> reference;
+    private final NessieConfig config;
+
+    public NessieIcebergClient(NessieApiV1 nessieApi, NessieConfig config)
+    {
+        this.nessieApi = requireNonNull(nessieApi, "nessieApi is null");
+        this.config = requireNonNull(config, "nessieConfig is null");
+        this.reference = () -> loadReference(Optional.ofNullable(config.getDefaultReferenceName()), null);
+    }
+
+    public UpdateableReference getReference()
```

Comment:
This method is used extensively and it makes every time when being called API requests to Nessie API.
Does it need to be this way?

Reply:
this is in fact an oversight. It should be using `Suppliers.memoize`